File: game/skills/LevelUp.lua
-- [ts]: LevelUp.ts
local ____lualib = require("lualib_bundle") -- 1
local __TS__Class = ____lualib.__TS__Class -- 1
local __TS__ArrayMap = ____lualib.__TS__ArrayMap -- 1
local __TS__SetDescriptor = ____lualib.__TS__SetDescriptor -- 1
local ____exports = {} -- 1
local ____GameContext = require("game.core.GameContext") -- 3
local ctx = ____GameContext.ctx -- 3
local ____Config = require("game.config.Config") -- 4
local Config = ____Config.Config -- 4
local ____SkillDefs = require("game.skills.SkillDefs") -- 5
local getSkillDef = ____SkillDefs.getSkillDef -- 5
local poolForCharacter = ____SkillDefs.poolForCharacter -- 5
____exports.LevelUpSystem = __TS__Class() -- 9
local LevelUpSystem = ____exports.LevelUpSystem -- 9
LevelUpSystem.name = "LevelUpSystem" -- 9
function LevelUpSystem.prototype.____constructor(self, skillSystem, rng) -- 19
	self.currentChoices = {} -- 12
	self.pendingLevels = 0 -- 13
	self.skillSystem = skillSystem -- 20
	self.rng = rng -- 21
	ctx.onPlayerLevelUp = function() return self:beginLevelUp() end -- 23
end -- 19
function LevelUpSystem.prototype.offerChoices(self, characterId) -- 32
	local pool = poolForCharacter(characterId, self.rng) -- 33
	local result = {} -- 34
	do -- 34
		local i = 0 -- 35
		while i < #pool and #result < Config.levelUpChoices do -- 35
			local id = pool[i + 1] -- 36
			if self.skillSystem:getLevel(id) < getSkillDef(id).maxStack then -- 36
				result[#result + 1] = id -- 38
			end -- 38
			i = i + 1 -- 35
		end -- 35
	end -- 35
	return result -- 41
end -- 32
function LevelUpSystem.prototype.beginLevelUp(self) -- 45
	if ctx.phase == "levelup" then -- 45
		self.pendingLevels = self.pendingLevels + 1 -- 47
		return -- 48
	end -- 48
	local ____opt_0 = ctx.character -- 48
	local charId = ____opt_0 and ____opt_0.id -- 50
	if charId == nil then -- 50
		self:finishLevelUp() -- 53
		return -- 54
	end -- 54
	self.currentChoices = self:offerChoices(charId) -- 56
	ctx.phase = "levelup" -- 57
	local ____opt_2 = self.onChoicesReady -- 57
	if ____opt_2 ~= nil then -- 57
		____opt_2( -- 58
			self, -- 58
			__TS__ArrayMap( -- 58
				self.currentChoices, -- 58
				function(____, id) return getSkillDef(id) end -- 58
			) -- 58
		) -- 58
	end -- 58
end -- 45
function LevelUpSystem.prototype.choose(self, index) -- 62
	local id = self.currentChoices[index + 1] -- 63
	if id == nil then -- 63
		return -- 64
	end -- 64
	self.skillSystem:applySkill(id) -- 65
	self:finishLevelUp() -- 66
end -- 62
function LevelUpSystem.prototype.finishLevelUp(self) -- 70
	self.currentChoices = {} -- 71
	if self.pendingLevels > 0 then -- 71
		self.pendingLevels = self.pendingLevels - 1 -- 73
		ctx.phase = "playing" -- 74
		self:beginLevelUp() -- 75
		return -- 76
	end -- 76
	ctx.phase = "playing" -- 78
	local ____opt_4 = self.onLevelUpFinished -- 78
	if ____opt_4 ~= nil then -- 78
		____opt_4(self) -- 79
	end -- 79
end -- 70
function LevelUpSystem.prototype.reset(self) -- 83
	self.currentChoices = {} -- 84
	self.pendingLevels = 0 -- 85
end -- 83
__TS__SetDescriptor( -- 83
	LevelUpSystem.prototype, -- 83
	"choices", -- 83
	{get = function(self) -- 83
		return __TS__ArrayMap( -- 28
			self.currentChoices, -- 28
			function(____, id) return getSkillDef(id) end -- 28
		) -- 28
	end}, -- 28
	true -- 28
) -- 28
return ____exports -- 28

File: game/skills/LevelUp.ts
// 升级三选一系统：玩家域触发 ctx.onPlayerLevelUp → 生成候选 → 面板展示 → 选择应用
import type { CharacterId, SkillDef, SkillId } from 'game/core/Types';
import { ctx } from 'game/core/GameContext';
import { Config } from 'game/config/Config';
import { getSkillDef, poolForCharacter } from 'game/skills/SkillDefs';
import type { SkillSystem } from 'game/skills/SkillSystem';
import type { RNG } from 'game/utils/RNG';

export class LevelUpSystem {
	private skillSystem: SkillSystem;
	private rng: RNG;
	private currentChoices: SkillId[] = [];
	private pendingLevels = 0;
	// UI 面板订阅：候选生成后回调（传 SkillDef[] 便于直接渲染）
	onChoicesReady?: (choices: SkillDef[]) => void;
	// 主控 Game.ts 订阅：三选一完成后通知恢复 phase='playing'
	onLevelUpFinished?: () => void;

	constructor(skillSystem: SkillSystem, rng: RNG) {
		this.skillSystem = skillSystem;
		this.rng = rng;
		// 玩家域在 Player.addExp 升级时调用 ctx.onPlayerLevelUp，这里接线
		ctx.onPlayerLevelUp = (): void => this.beginLevelUp();
	}

	// 当前候选技能定义（供面板查询）
	get choices(): SkillDef[] {
		return this.currentChoices.map((id) => getSkillDef(id));
	}

	// 从角色候选池抽取 Config.levelUpChoices 个未满级技能
	offerChoices(characterId: CharacterId): SkillId[] {
		const pool = poolForCharacter(characterId, this.rng);
		const result: SkillId[] = [];
		for (let i = 0; i < pool.length && result.length < Config.levelUpChoices; i++) {
			const id = pool[i];
			if (this.skillSystem.getLevel(id) < getSkillDef(id).maxStack) {
				result.push(id);
			}
		}
		return result;
	}

	// 进入升级选择：生成三选一 → 切 phase → 通知面板
	beginLevelUp(): void {
		if (ctx.phase === 'levelup') {
			this.pendingLevels++;
			return;
		}
		const charId = ctx.character?.id;
		if (charId === undefined) {
			// 角色未设置时无法选池，直接结束避免卡死
			this.finishLevelUp();
			return;
		}
		this.currentChoices = this.offerChoices(charId);
		ctx.phase = 'levelup';
		this.onChoicesReady?.(this.currentChoices.map((id) => getSkillDef(id)));
	}

	// 选择第 index 项：应用技能 → 恢复游玩
	choose(index: number): void {
		const id = this.currentChoices[index];
		if (id === undefined) return;
		this.skillSystem.applySkill(id);
		this.finishLevelUp();
	}

	// 结束升级选择：恢复游玩阶段（主控可经 onLevelUpFinished 再接线）
	finishLevelUp(): void {
		this.currentChoices = [];
		if (this.pendingLevels > 0) {
			this.pendingLevels--;
			ctx.phase = 'playing';
			this.beginLevelUp();
			return;
		}
		ctx.phase = 'playing';
		this.onLevelUpFinished?.();
	}

	// 新对局重置
	reset(): void {
		this.currentChoices = [];
		this.pendingLevels = 0;
	}
}
